src/kaleidoscope/MatrixAddr.h

```diff
@@ -0,0 +1,385 @@
+  MatrixAddr &operator+(const MatrixAddr__ &other) {
+    this->row_ += other.row();
+    this->col_ += other.col();
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator-(const MatrixAddr__ &other) {
+    this->row_ -= other.row();
+    this->col_ -= other.col();
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator+=(const MatrixAddr__ &other) {
+    *this = *this + other;
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator-=(const MatrixAddr__ &other) {
+    *this = *this - other;
+    return *this;
+  }
+};
+
+// This is a copressed storage version of MatrixAddr.
+// It uses a common integer to store row and column indices.
+// To enable fast row/col access, bitshifts are used instead of
+// integer divisions to compute row indices. This is only possible
+// by padding rows with unused entries to gain a row size that
+// is a power of two.
+//
+template<uint8_t rows__, uint8_t cols__, typename StorageType__ = uint8_t>
+class MatrixAddrCompressed {
+ private:
+
+  StorageType__ raw_;
+
+ public:
+
+  typedef MatrixAddrCompressed<rows__, cols__, StorageType__> ThisType;
+
+  // Row size must be a power of two to enable fast index extraction.
+  static constexpr uint8_t rowSize
+    = internal::matrix_addr::upperPowerOfTwo(cols__);
+
+  static constexpr StorageType__ maxRawValue
+    = internal::matrix_addr::IntLimitsTrait<StorageType__>::max;
+  static constexpr StorageType__ invalidState = maxRawValue;
+  static constexpr uint8_t shift = internal::matrix_addr::log2(rowSize);
+
+  static constexpr uint8_t rows = rows__;
+  static constexpr uint8_t cols = cols__;
+
+  static constexpr uint8_t matrixSize = rows__ * cols__;
+
+  static constexpr StorageType__ rowMask = StorageType__(maxRawValue << shift);
+  static constexpr StorageType__ colMask = StorageType__(~rowMask);
+
+  static_assert(rows * rowSize < maxRawValue, "Please use a StorageType__ with a higher capacity");
+
+  constexpr MatrixAddrCompressed() : raw_(invalidState) {}
+
+  explicit constexpr MatrixAddrCompressed(uint8_t row, uint8_t col) : raw_(row * rowSize + col) {}
+
+  explicit constexpr MatrixAddrCompressed(uint8_t offset)
+    : MatrixAddrCompressed(offset / cols, offset % cols) {}
+
+  constexpr MatrixAddrCompressed(const ThisType &other) : raw_(other.raw_) {}
+
+  template<typename MatrixAddr__>
+  constexpr MatrixAddrCompressed(const MatrixAddr__ &other)
+    : MatrixAddrCompressed(other.row(), other.col()) {}
+
+  constexpr uint8_t row() const {
+    return raw_ >> shift;
+  }
+  constexpr uint8_t col() const {
+    return raw_ & colMask;
+  }
+
+  void setRow(uint8_t r) {
+    raw_ = (r << shift) | (raw_ & colMask);
+  }
+  void setCol(uint8_t c) {
+    raw_ = (raw_ & rowMask) | c;
+  }
+
+  constexpr uint8_t offset() {
+    return row() * cols + col();
+  }
+
+  constexpr bool isValid() const {
+    return (row() < rows__) && (col() < cols__);
```

Comment:
Why continue basing everything on row and column here? Why not just use a single number for the first invalid key index? Also, what if a keyboard with holes in the matrix has some reason to guarantee that `isValid()` will only return true for matrix addresses that correspond to actual keyswitches?

Reply:
I have no intention to move away from row/column indexing with this PR. Let's keep things separate.

`isValid` could be easily be renamed `isInRange`. In no placer does it replace anything related to holes.

Reply:
Some thoughts about linear indexing: 

There are many places in the firmware where linear indexing instead of two dimensional indexing would likely be more efficient. Nested for loops could be merged into one loop and index computations for lookup in two dimensional arrays could be saved. I think I have even seen places where a row/col pair is converted to a linear index and then back to row/col before it is actually used. That's probably what you have in mind with your linear design.

But this is a keyboard firmware and AFAIK most keyboards come with some sort of matrix hardware. That's why row/col indices make some sense to me. Yet I am not sure if they need to be exposed to the user as part of interfaces. Dealing with row/col indexing could as well be restricted to the hardware classes and the keyscanner code.

Most keyboards - except for some special designs like the Planck - have a weird indexing scheme anyway that makes it impossible for the user to remember the row/col values of specific keys. On such devices linear indexing would not be worse than row/col indexing because when a specific key s supposed to be addressed, the user must look it up in [some sort of table](https://github.com/keyboardio/Kaleidoscope-Hardware-Model01/blob/f469015346535cb864a340bf8eb317d268943248/src/Kaleidoscope-Hardware-Model01.h#L267-L279) anyway.

Nevertheless, I favor evolution over revolution and that's why I would like to see small changes - and this PR is a first step. Based on its changes it will be easier to switch to a linear indexing if the core designers would agree on that.